lib/experimental/media/class-gutenberg-rest-attachments-controller.php

```diff
@@ -0,0 +1,361 @@
+
+			$matches = array();
+			if ( preg_match( '/(.*)(-\d+x\d+)-' . $number . '$/', $name, $matches ) ) {
+				$filename_without_suffix = $matches[1] . $matches[2] . ".$ext";
+				if ( $matches[1] === $orig_name && ! file_exists( "$dir/$filename_without_suffix" ) ) {
+					return $filename_without_suffix;
+				}
+			}
+
+			return $filename;
+		};
+
+		add_filter( 'wp_unique_filename', $filter_filename, 10, 6 );
+
+		// See https://github.com/swissspidy/media-experiments/issues/465.
+		// See https://core.trac.wordpress.org/ticket/61189.
+		if ( version_compare( get_bloginfo( 'version' ), '6.6-beta1', '>=' ) ) {
+			$parent_post = get_post_parent( $attachment_id );
+
+			$time = null;
+
+			// Matches logic in media_handle_upload().
+			// The post date doesn't usually matter for pages, so don't backdate this upload.
+			if ( $parent_post && 'page' !== $parent_post->post_type && substr( $parent_post->post_date, 0, 4 ) > 0 ) {
+				$time = $parent_post->post_date;
+			}
+
+			if ( ! empty( $files ) ) {
+				$file = $this->upload_from_file( $files, $headers, $time );
+			} else {
+				$file = $this->upload_from_data( $request->get_body(), $headers, $time );
+			}
+		} else {
+			if ( ! empty( $files ) ) {
+				$file = $this->upload_from_file( $files, $headers );
+			} else {
+				$file = $this->upload_from_data( $request->get_body(), $headers );
+			}
+		}
+
+		remove_filter( 'wp_unique_filename', $filter_filename );
+		remove_filter( 'image_editor_output_format', '__return_empty_array', 100 );
+
+		if ( is_wp_error( $file ) ) {
+			return $file;
+		}
+
+		$type = $file['type'];
+		$path = $file['file'];
+
+		// TODO: Better fallback if image_size is not provided.
+		$image_size = $request['image_size'];
+
+		$metadata = wp_get_attachment_metadata( $attachment_id, true );
+
+		if ( ! $metadata ) {
+			$metadata = array();
+		}
+
+		if ( 'original' === $image_size ) {
+			$metadata['original_image'] = basename( $path );
+		} else {
+			$metadata['sizes'] = $metadata['sizes'] ?? array();
+
+			$size = wp_getimagesize( $path );
+
+			$metadata['sizes'][ $image_size ] = array(
+				'width'     => $size ? $size[0] : 0,
+				'height'    => $size ? $size[1] : 0,
+				'file'      => basename( $path ),
+				'mime-type' => $type,
+				'filesize'  => wp_filesize( $path ),
+			);
+		}
+
+		wp_update_attachment_metadata( $attachment_id, $metadata );
+
+		$response_request = new WP_REST_Request(
+			WP_REST_Server::READABLE,
+			rest_url( sprintf( '%s/%s/%d', $this->namespace, $this->rest_base, $attachment_id ) )
+		);
+
+		$response_request->set_param( 'context', 'edit' );
+
+		if ( isset( $request['_fields'] ) ) {
+			$response_request['_fields'] = $request['_fields'];
+		}
+
+		$response = $this->prepare_item_for_response( get_post( $attachment_id ), $response_request );
+
+		$response->set_status( 201 );
+		$response->header( 'Location', rest_url( sprintf( '%s/%s/%d', $this->namespace, $this->rest_base, $attachment_id ) ) );
+
+		return $response;
+	}
+}
```

Comment:
Do you think all the endpoint changes can be extracted to a separate PR to ease reviews... We can ping folks working on the REST API to help.

Reply:
Sure yeah that's possible. It won't be useful on its own, but makes reviewing easier for sure.